### ISSUE TITLE
Design proposals

### ISSUE THREAD

**Issue by HDembinski** [impy-project/impy#3]
One of the fundamental principles in good software design is to use the most basic language features to implement some feature. In other words, be minimalist when designing the interface. Always ask yourself: is there a simpler more basic way of providing this functionality? Another principle is: "Make interfaces easy to use correctly and hard or impossible to use incorrectly." This is obvious, but you have to really ask yourself this question every time you design something. How can I make it so that it is easy to use correctly and hard or impossible to use incorrectly?

Let's discuss the example code from that perspective.

```py
from impy.definitions import *
from impy.constants import *
from impy.kinematics import EventKinematics
from impy import impy_config

# Define the parameters of the collisions
event_kinematics = EventKinematics(
    ecm=13 * TeV, p1pdg=2212, p2pdg=2212)

# Create an instance of an event generator by passing
# the model name as a string
generator = make_generator_instance(
    interaction_model_by_tag['SIBYLL2.3D'])

# Initialize it
generator.init_generator(event_kinematics)

# Number of events to generate
nevents = 100

for event in generator.event_generator(event_kinematics, nevents):
    event.filter_final_state_charged()
    # do something with event.p_ids, event.eta, event.en, event.pt, etc.
    # these variables are numpy arrays, that can be histogrammed or counted like
    average_pt += 1/float(nevents)*np.mean(event.pt[np.abs(event.p_ids) == 211])

print('Average pT for charged pions {0:4.3f}'.format(average_pt))
```

* `EventKinematics` seems ok, it is probably useful to bundle this in a class. Providing units like `TeV` is good. Ideally, those are not just numbers but some class with `__mul__` and `__rmul__` operators defined, so that the `EventKinematics` object can check that the user actually supplied the unit and give an error when a plain number is passed to it. This is a bit fancy, so has low priority, but it would be the right way to implement this (make it impossible to use it incorrectly).

* `generator = make_generator_instance(interaction_model_by_tag['SIBYLL2.3D'])`. This does not seem minimalist. `make_generator_instance` seems clunky, and why do I need the dict `interaction_model_by_tag` as well? If you need a generator factory, why can't this factory understand the model string 'SIBYLL2.3D' directly? I think there should be just one way to grab the interaction model, and the most basic way is to import each model as a class. If there is some common code that needs to be executed for each model, then this code should go into the init of the base class from which each model class is derived. In Python there is no reason to use a factory function to create objects. So the ideal syntax from my PoV would be `from model import sibyll_23d; generator = sibyll_23d(event_kinematics)`. There are many advantages to this, notably, **smart editors can provide autocompletion for the generator name in this case**, which is impossible if the generator name is a string passed to the generator factory.

* In OOP, there is a rule: objects should be fully initialized in a working state after the ctor. This means, you should get rid of `init_generator`, this should be called inside the `__init__` of the generator when you create the object from the class. It is ok to make it so that you can reuse the object by assigning a different `event_kinematics`, in that case you would make `event_kinematics` into a property of the generator instance and allow assigning to it. Internally, this can call `init_generator` then which should be a private method starting with `_`

* `for event in generator.event_generator(event_kinematics, nevents):` this should just be `for event in generator(nevents):` You already passed `event_kinematics` to the generator, you should not have to do it again.

In summary, here is how this code should look like with these changes implemented:

```py
from impy.models import sibyll_23d
from impy.constants import TeV
from impy.kinematics import EventKinematics

event_kinematics = EventKinematics(
    ecm=13 * TeV, p1pdg=2212, p2pdg=2212)

generator = sibyll_23d(event_kinematics)

# Number of events to generate
nevents = 100

for event in generator(nevents):
    event.filter_final_state_charged() # I don't like this, but let's discuss it elsewhere
    average_pt += 1/float(nevents)*np.mean(event.pt[np.abs(event.p_ids) == 211])
```

**Comment by HDembinski** [comment]
A separate topic. Should event filters be members of the event class or separate functions?
`event.filter_final_state_charged()`

Making them members of the event class only makes sense if you want to filter in-place, as this is more efficient and avoids allocating another array. If that is the goal, then the filter should return the event object, so that one can chain several filter calls together  `event.filter_A().filter_B().filter_C()`. I am not sure, perhaps that is already the case.

**Comment by HDembinski** [comment]
Another aspect of good software design is to be very careful what parts of the API are exposed to the user. As you gain users and they start to rely on your API, you will find it difficult to change the API. So the less interface you expose publicly, the better, as this allows you to make changes without breaking user code.

As a consequence, this again leads to minimal APIs. You should make everything private (starting with `_`) unless it is absolutely necessary for a user to be able to call this.

**Comment by afedynitch** [comment]
I agree with these points.

- The units thing is fancy, indeed. The units are only used for the `EventKinematics` objects.
- The generators was actually your proposal from 2-3 years ago :) I had classes but it was clunky. Can't your proposal happen almost entirely in an `__init__.py`.
- Calling `init` in the constructer should be possible now, indeed, and I don't see any reason why it shouldn't work. There were some historical reasons for that. Also, note, the classes actually have to be singleton since their fortran libraries can not be "unloaded" in runtime.
-  Agree `.event_generator` is also not necessary

I definitely like the way how the new code looks like! These suggestions can all become issues.

Regarding the filters: as it is now the filters change the views to the numpy arrays inplace for the entire event. One can do something like
```
for event in generator(nevents):
    # Do something to the entire particle stack, like asking for mothers and daughters, generate a hepmc event
    event.filter_final_state_charged() # Tighten the view
    # Do something on a reduced particle list
    average_pt += 1/float(nevents)*np.mean(event.pt[np.abs(event.p_ids) == 211])
```
I don't really imagine how `event.filter_A().filter_B()` should work in practice? Maybe something like
```
for event in generator(nevents):
    evt = event.filter_final().filter_charged().filter_strange()
    # Do something 
```

**Comment by HDembinski** [comment]
> I don't really imagine how `event.filter_A().filter_B()` should work in practice? Maybe something like
> 
> ```
> for event in generator(nevents):
>     evt = event.filter_final().filter_charged().filter_strange()
>     # Do something 
> ```

Almost.

```py
for event in generator(nevents):
     event.filter_final().filter_charged().filter_strange()
     # Do something with event
```

**Comment by HDembinski** [comment]
>     * Calling `init` in the constructor should be possible now, indeed, and I don't see any reason why it shouldn't work. There were some historical reasons for that. Also, note, the classes actually have to be singleton since their fortran libraries can not be "unloaded" in runtime.

Ok, I see, I forgot about the singleton issue. Then classes are not the perfect model, but I think the advantages of using classes still outweigh the issues. There has to be a big warning though that one cannot use generators with Python's thread-based multiprocessing, only with Python process-based multiprocessing.
